airbyte-workers/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java

```diff
@@ -306,31 +306,11 @@ private static Runnable getReplicationRunnable(final AirbyteSource source,
           } catch (final Exception e) {
             throw new SourceException("Source process read attempt failed", e);
           }
-          if (messageOptional.isPresent()) {
-            if (messageOptional.get().getRecord() != null) {
-              final AirbyteRecordMessage message = messageOptional.get().getRecord();
-              final String messageStream = WorkerUtils.streamNameWithNamespace(message.getNamespace(), message.getStream());
-              // validate a record's schema if there are less than 10 records with validation errors
-              if (validationErrors.get(messageStream) == null || validationErrors.get(messageStream).getRight() < 10) {
-                try {
-                  recordSchemaValidator.validateSchema(messageOptional.get().getRecord(), messageStream);
-                } catch (final RecordSchemaValidationException e) {
-                  final ImmutablePair<Set<String>, Integer> exceptionWithCount = validationErrors.get(messageStream);
-                  if (exceptionWithCount == null) {
-                    validationErrors.put(messageStream, new ImmutablePair<>(e.errorMessages, 1));
-                  } else {
-                    final Integer currentCount = exceptionWithCount.getRight();
-                    final Set<String> currentErrorMessages = exceptionWithCount.getLeft();
-                    final Set<String> updatedErrorMessages =
-                        Stream.concat(currentErrorMessages.stream(), e.errorMessages.stream()).collect(Collectors.toSet());
-                    validationErrors.put(messageStream, new ImmutablePair<>(updatedErrorMessages, currentCount + 1));
-                  }
-                }
-
-              }
-            }
 
-            final AirbyteMessage message = mapper.mapMessage(messageOptional.get());
+          if (messageOptional.isPresent()) {
+            final AirbyteMessage airbyteMessage = messageOptional.get();
+            validateSchema(recordSchemaValidator, validationErrors, airbyteMessage);
```

Comment:
pull all this logic into a separate function to keep logic at the same detail layers e.g. the top level replication function should only know about high-level replication related code.